pkg/etcd/etcd.go

```diff
@@ -963,6 +956,104 @@ func (e *ETCD) listSnapshots(ctx context.Context, snapshotDir string) ([]snapsho
 	return snapshots, nil
 }
 
+// initS3IfNil initializes the S3 client
+// if it hasn't yet been initialized.
+func (e *ETCD) initS3IfNil(ctx context.Context) error {
+	if e.s3 == nil {
+		s3, err := newS3(ctx, e.config)
+		if err != nil {
+			return err
+		}
+		e.s3 = s3
+	}
+
+	return nil
+}
+
+// deleteSnapshots removes the given snapshots from
+// either local storage or S3.
+func (e *ETCD) DeleteSnapshots(ctx context.Context, snapshots []string) error {
+	snapshotDir, err := snapshotDir(e.config)
+	if err != nil {
+		return errors.Wrap(err, "failed to get the snapshot dir")
+	}
+
+	if e.config.EtcdS3 {
+		logrus.Info("Removing the given etcd snapshot from S3")
+		logrus.Debugf("Removing the given etcd snapshot from S3: %v", snapshots)
+
+		if e.initS3IfNil(ctx); err != nil {
+			return err
+		}
+
+		if len(snapshots) > 1 {
+			logrus.Info("Removing the given etcd snapshot(s) from S3")
+
+			objectsCh := make(chan minio.ObjectInfo)
+
+			toCtx, cancel := context.WithTimeout(ctx, time.Second*30)
+			defer cancel()
+
+			go func(ctx context.Context) {
+				defer close(objectsCh)
+
+				for {
```

Comment:
Lacking a timer `case` for the `select` is the `for` necessary?

Reply:
Actually, the `for` + `select` is wholly unnecessary as you are passing down the `ctx` that was passed into the goroutine

Reply:
The idea is to enforce a timeout on the ListObjects call so that this does not block excessively; the outer ctx is just the main application context that's tied to the SIGINT/SIGTERM handlers and will effectively never be cancelled until the whole process exits.

The for/select here will ensure that a 30 second timeout is enforced for both individual ListObjects calls, and prevent additional calls from occurring if the context has been cancelled.

Reply:
> @dweomer the idea is to enforce a timeout on the ListObjects call so that this does not block excessively; the outer ctx is just the main application context that's tied to the SIGINT/SIGTERM handlers and will effectively never be cancelled until the whole process exits.
> 
> The for/select here will ensure that a 30 second timeout is enforced for both individual ListObjects calls, and prevent additional calls from occurring once the time limit has been exceeded.

I don't think what is written works the way you've described it. There is no `select` case against a `time.After(<short duration>)` + `continue` which is what I would expect to see in such a construction. As it is written, the code will exit early if the context is cancelled or it will make the single call to `e.s3.client.ListObjects` and potentially hang up to the context's timeout. There is no control flow that would cause the "loop" to run through more than a single pass.